### PR TITLE
Update generator-langium web template

### LINKED ISSUE #1163: Browser template should be updated
As soon as `monaco-editor-wrapper` v3 becomes available the browser template of the generator should be updated. Let's jump over v2.

### PR DESCRIPTION
**Pull request by kaisalmen**
~**We need to wait for `monaco-editor-wrapper@3.3.0` ([PR54](https://github.com/TypeFox/monaco-components/pull/54)) before we should merge this change.**~ (merged)

This PR fixes #1163 and updates the web template to the latest `monaco-editor-wrapper`.
It is now possible to use the textmate grammar definition as well as the monarch grammar definitions with `monaco-editor`. The underlying [monaco-vscode-api](https://github.com/CodinGame/monaco-vscode-api) exposes various vscode service overrides that make this possible. Therefore we now generate two HTML examples and a top-level index page.

The generated output no longer requires express. vite+rollup are used to produce the bundles, so the output can be used/served as is. After the generation/template building is complete the output can be tested using vite's dev server.

### CODE REVIEW

File: packages/generator-langium/templates/web/.package.json
Comment (montymxb) [inline packages/generator-langium/templates/web/.package.json:7]: This didn't work when I tested things out locally. The main browser is still in `./out/language/main-browser.js`, but this was the only hiccup I found while testing out the generator.

**Review by montymxb — CHANGES_REQUESTED** [review]
One apparent bug with the generated package.json, and a few comments, but nothing else that I could find. Seems like the rest is working as intended.

File: packages/generator-langium/templates/web/.package.json
Comment (kaisalmen) [reply packages/generator-langium/templates/web/.package.json:7]: This is weird. Both Windows and Linux produce the dir structure as seen in the screenshot and therefore I don't see that error. I changed the command, because I had the opposite problem 🤷‍♂️ Did you use `npm run run`? 

image

File: packages/generator-langium/templates/web/vite.bundle.classic.config.ts
Comment (kaisalmen) [reply packages/generator-langium/templates/web/vite.bundle.classic.config.ts:19]: Good point. I will change that.

File: packages/generator-langium/templates/web/vite.bundle.vscodeApi.config.ts
Comment (kaisalmen) [reply packages/generator-langium/templates/web/vite.bundle.vscodeApi.config.ts:19]: Good point. I will change that.

File: packages/generator-langium/templates/web/scripts/bundleVscodeApi.js
Comment (kaisalmen) [reply packages/generator-langium/templates/web/scripts/bundleVscodeApi.js]: Thanks for the hint. It can be just that (fixed in next commit ⬇️):
```javascript
import 'vscode/default-extensions/theme-defaults';
export * from './bundleClassic.js';
```

File: packages/generator-langium/templates/web/.package.json
Comment (montymxb) [reply packages/generator-langium/templates/web/.package.json:7]: Yep, was working through `npm run run`, and post-clean as well. Possible difference due to platforms?

Comment (kaisalmen) [reply packages/generator-langium/templates/web/.package.json]: Fixed in next commit.

**Comment by kaisalmen** [comment]
@montymxb code is updated to final `monaco-editor-wrapper@3.2.0`

**Comment by kaisalmen** [comment]
Rebased to latest main and updated to latest `monaco-editor-wrapper`.

**Comment by Yokozuna59** [comment]
Hi @kaisalmen, It may be a silly question, but what's the difference between `classic` and `vscodeApi`? And is there a way to test it out before the next release?

**Comment by kaisalmen** [comment]
@Yokozuna59 it is basically this:

- **classic**: Configure `monaco-editor` as you would when using it directly
- **Extended**: Configure `monaco-editor` like a VSCode extension

The extended mode allows all the fancy things from: [monaco-vscode-api](https://github.com/CodinGame/monaco-vscode-api)

There is now another PR in the pipeline for the wrapper that makes the mode separation clearer (terminology and code): https://github.com/TypeFox/monaco-components/pull/54

You can checkout the PR locally and use the generator to try it out. It generates both versions. You can navigate from the index page:
image

**Comment by Yokozuna59** [comment]
@kaisalmen I see, thanks for illustration.

> There is now another PR in the pipeline for the wrapper that makes the mode separation clearer (terminology and code): [TypeFox/monaco-components#54](https://github.com/TypeFox/monaco-components/pull/54)

Is there an estimated time when it will be merged? After looking at the changes in that PR, it contains a lot of breaking changes, and it would be hard to reimplement the code to inline it with the next major release.

**Comment by kaisalmen** [comment]
> After looking at the changes in that PR, it contains a lot of breaking changes, and it would be hard to reimplement the code to inline it with the next major release.

@Yokozuna59 the changes in `monaco-editor-wrapper` are mostly internal and don't change the external API much. The only change required here is to change from $type from `vscode` to `extended` apart from updating the version. next versions are available on npm (see description of [#54](https://github.com/TypeFox/monaco-components/pull/54))

**Comment by Yokozuna59** [comment]
Hi @kaisalmen, just a simple question, is there a reason why `vscode-languageserver` isn't a dependency in the template? Both web and vscode templates uses it:

- Web template:

    https://github.com/eclipse-langium/langium/blob/0708200d01ccd52880a992dae349e080f1e99c38/packages/generator-langium/templates/web/src/language/main-browser.ts#L2

- VSCode template:

    https://github.com/eclipse-langium/langium/blob/0708200d01ccd52880a992dae349e080f1e99c38/packages/generator-langium/templates/vscode/src/language/main.ts#L3

I asked that because when using `npm` package manager, it works fine and build without issues. But I switched to `pnpm` it started throwing errors:

```
> hello-world@0.0.1 build hello-world
> tsc -b tsconfig.json && node esbuild.mjs

src/language/main-browser.ts:6:8 - error TS2307: Cannot find module 'vscode-languageserver/browser.js' or its corresponding type declarations.

6 } from 'vscode-languageserver/browser.js';
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/language/main.ts:3:52 - error TS2307: Cannot find module 'vscode-languageserver/node.js' or its corresponding type declarations.

3 import { createConnection, ProposedFeatures } from 'vscode-languageserver/node.js';
                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.

 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by Yokozuna59** [comment]
I was able to resolve some of the errors that were thrown when switching to `pnpm`:

1. install missing dependencies:

    ```bash
    pnpm add monaco-languageclient@6.5.2 \
      vscode-languageserver@8.0.2 \
      vscode@npm:@codingame/monaco-vscode-api@1.82.4
    ```

> **Note:** installing `@codingame/monaco-vscode-api` is only to resolve the `peerDependencies` missing issue.

2. Make `vscode` as external dependency:

```ts
// vite.bundle.classic.config.ts
export default defineConfig({
    build: {
        ...,
        rollupOptions: {
            external: ['vscode'],
            ...,
        }
    }
});
```

These changes don't have any impact when using `npm`.

---

But I don't know how to resolve other issues:


<details>

```
> hello-world@0.0.1 build:bundles
> vite --config ./vite.bundle.classic.config.ts build && vite --config ./vite.bundle.vscodeApi.config.ts build

vite v4.4.11 building for production...
✓ 8 modules transformed.
✓ built in 184ms
[vite]: Rollup failed to resolve import "monaco-editor/esm/vs/platform/userDataProfile/common/userDataProfile.js" from "hello-world/node_modules/.pnpm/@codingame+monaco-vscode-api@1.82.4/node_modules/@codingame/monaco-vscode-api/service-override/configuration.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
error during build:
Error: [vite]: Rollup failed to resolve import "monaco-editor/esm/vs/platform/userDataProfile/common/userDataProfile.js" from "hello-world/node_modules/.pnpm/@codingame+monaco-vscode-api@1.82.4/node_modules/@codingame/monaco-vscode-api/service-override/configuration.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
    at viteWarn (file://hello-world/node_modules/.pnpm/vite@4.4.11_@types+node@20.8.4/node_modules/vite/dist/node/chunks/dep-2b82a1ce.js:48205:27)
    at onRollupWarning (file://hello-world/node_modules/.pnpm/vite@4.4.11_@types+node@20.8.4/node_modules/vite/dist/node/chunks/dep-2b82a1ce.js:48237:9)
    at onwarn (file://hello-world/node_modules/.pnpm/vite@4.4.11_@types+node@20.8.4/node_modules/vite/dist/node/chunks/dep-2b82a1ce.js:47965:13)
    at file://hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:24276:13
    at Object.logger [as onLog] (file://hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:25950:9)
    at ModuleLoader.handleInvalidResolvedId (file://hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:24862:26)
    at file://hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:24822:26
 ELIFECYCLE  Command failed with exit code 1.
```

</details>

And this issue occurs for different imports all over `@codingame/monaco-vscode-api`. I sat all those files as external as workaround, but another issue was thrown:

<details>

```
> hello-world@0.0.1 build:bundles
> vite --config ./vite.bundle.classic.config.ts build && vite --config ./vite.bundle.vscodeApi.config.ts build

vite v4.4.11 building for production...
✓ 2044 modules transformed.
✓ built in 8.85s
"FileType" is not exported by "node_modules/.pnpm/monaco-editor@0.43.0/node_modules/monaco-editor/esm/vs/platform/files/common/files.js", imported by "node_modules/.pnpm/@codingame+monaco-vscode-api@1.82.4/node_modules/@codingame/monaco-vscode-api/vscode/src/vs/workbench/common/editor.js".
file: /hello-world/node_modules/.pnpm/@codingame+monaco-vscode-api@1.82.4/node_modules/@codingame/monaco-vscode-api/vscode/src/vs/workbench/common/editor.js:7:9
5: import { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
6: import { Registry } from 'monaco-editor/esm/vs/platform/registry/common/platform.js';
7: import { FileType } from 'monaco-editor/esm/vs/platform/files/common/files.js';
            ^
8: import { Schemas } from 'monaco-editor/esm/vs/base/common/network.js';
9: import { isErrorWithActions, createErrorWithActions } from 'monaco-editor/esm/vs/base/common/errorMessage.js';
error during build:
RollupError: "FileType" is not exported by "node_modules/.pnpm/monaco-editor@0.43.0/node_modules/monaco-editor/esm/vs/platform/files/common/files.js", imported by "node_modules/.pnpm/@codingame+monaco-vscode-api@1.82.4/node_modules/@codingame/monaco-vscode-api/vscode/src/vs/workbench/common/editor.js".
    at error (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:2287:30)
    at Module.error (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:13745:16)
    at Module.traceVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:14175:29)
    at ModuleScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:12615:39)
    at FunctionScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
    at ChildScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
    at ReturnValueScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
    at ChildScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
    at TrackingScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
    at BlockScope.findVariable (file:///hello-world/node_modules/.pnpm/rollup@3.29.4/node_modules/rollup/dist/es/shared/node-entry.js:7124:38)
 ELIFECYCLE  Command failed with exit code 1
```

</details>

I also tried to install `monaco-editor@0.43.0`, but it has no effect.

File: packages/generator-langium/templates/web/scripts/bundleClassic.js
Comment (Yokozuna59) [inline packages/generator-langium/templates/web/scripts/bundleClassic.js:7]: I'm not really sure why, but the types here is using [`css.d.ts`](https://github.com/TypeFox/monaco-components/blob/d7880556c4d69437a4ccd4ee071d48f4c21d32b0/packages/monaco-editor-wrapper/styles/css.d.ts) instead of [`index.d.ts`](https://github.com/TypeFox/monaco-components/blob/d7880556c4d69437a4ccd4ee071d48f4c21d32b0/packages/monaco-editor-wrapper/styles/index.d.ts):

image

After using the editor to see types, it opens this file:

image

**Comment by kaisalmen** [comment]
@Yokozuna59 thank you very much for your feedback. I will answer your feedback soon.
It is already clead to me that this PR should wait for `monaco-editor-wrapper@3.3.0` as it may already solve some of your issues and also it addresses some shortcoming I identified myself.

**Comment by kaisalmen** [comment]
> vscode-languageserver isn't a dependency in the template

You are correct. It works with npm, because it is a transitive dependency of langium, but it should be declared here.

File: packages/generator-langium/templates/web/tsconfig.monarch.json
Comment (kaisalmen) [reply packages/generator-langium/templates/web/tsconfig.monarch.json:17]: I will align this back to just 'static' everywhere in the next commit.

Comment (kaisalmen) [reply packages/generator-langium/templates/web/tsconfig.monarch.json]: Done, see below ⬇️

**Comment by Yokozuna59** [comment]
@kaisalmen Thanks for your fast response.

I guess the template `.gitignore` file has missing attributes.

```gitignore
static/bundle/
static/monaco-editor-workers/
static/worker/
```

File: packages/generator-langium/templates/web/scripts/bundleClassic.js
Comment (kaisalmen) [reply packages/generator-langium/templates/web/scripts/bundleClassic.js]: `typesVersions` were incorrectly specified. Fixed in latest commit/force-push.

**Comment by kaisalmen** [comment]
@montymxb this ready now. It is based on the final version `3.3.0` of `monaco-editor-wrapper`
@Yokozuna59 This version is now fully working with pnpm out of the box.

**Comment by kaisalmen** [comment]
@msujew this is ready. It needs approval from a core maintainer. 🙂

**Comment by kaisalmen** [comment]
@msujew the rebase updates `vscode-languageserver` and `vscode-languageclient` libraries to `9.0.1`.

Important: This will only work properly with a new langium release or a locally linked version. If you just check out and build it will fail, because Langium 2.0.x references another version that breaks the compilation.

**Comment by spoenemann** [comment]
> It needs approval from a core maintainer. 🙂

All committers listed here are allowed to approve and merge PRs:
https://projects.eclipse.org/projects/ecd.langium/who

**Review by montymxb — APPROVED** [review]
:octocat: approved! Generated project is good (although until the new langium release comes out, the generator will have a small mismatch).

**Pull request merged**
